### PR TITLE
Present claimed packages as projects

### PR DESCRIPTION
**Pull request by whit537**
Part of #4305, follows on #4397.

#### Todo

- [x] redirect `/on/npm/bar/` to `/deadbeef/` when claimed
- [x] fix regression in project receiving charts: https://github.com/gratipay/gratipay.com/pull/4398#issuecomment-293363521 —done in 582af24b08d05a9691e7ff78f28f874c07a13fed

### CODE REVIEW

**Comment by whit537** [comment]
screen shot 2017-04-04 at 4 06 22 pm

**Comment by whit537** [comment]
3347ae337c161ee8e52e975b33222f358fc75810 😍 

screen shot 2017-04-04 at 4 16 00 pm

**Comment by whit537** [comment]
Yesterday I started down a path of using Aspen typecasters to DRY up in a more general way. PR forthcoming against master for typecasting implementation, blocking this on that.

**Comment by whit537** [comment]
Alright, so we're looking at either implementing proper internal redirects, or serving package-projects on the same paths as regular projects after all.

The former is proving tricky (we either greatly complicate the `website.algorithm`, or we figure out how to rerun the entire algorithm). The latter should be fairly straightforward, though I guess that means further crowding our global project namespace w/ all of the npm namespace. Puts more pressure on #4246.

**Comment by whit537** [comment]
I guess that means we need to try to name projects after the underlying package on-claim, falling back as with usernames.

**Comment by whit537** [comment]
Bumped remaining todos to #4404. Ready for review!

**Comment by whit537** [comment]
c34da4ffebf1c70de04f2efe0acbb5f944959d8b or thereabouts breaks `www/%team/chart.json.spt`. Write a test for that and review other cases as well.

```
pid-89470 thread-140735094956800 (MainThread) Traceback (most recent call last):
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/algorithm.py", line 288, in run
pid-89470 thread-140735094956800 (MainThread)     new_state = function(**deps.as_kwargs)
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/aspen/algorithms/website.py", line 115, in get_response_for_resource
pid-89470 thread-140735094956800 (MainThread)     return {'response': resource.respond(state)}
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/aspen/http/resource.py", line 59, in respond
pid-89470 thread-140735094956800 (MainThread)     content_type, body = super(Dynamic, self).respond(accept, state)
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/aspen/simplates/__init__.py", line 169, in respond
pid-89470 thread-140735094956800 (MainThread)     exec(self.pages[1], context)
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/www/%team/charts.json.spt", line 48, in <module>
pid-89470 thread-140735094956800 (MainThread)     """, (slug,), back_as=dict)
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/__init__.py", line 548, in all
pid-89470 thread-140735094956800 (MainThread)     return cursor.all(sql, parameters)
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/cursors.py", line 145, in all
pid-89470 thread-140735094956800 (MainThread)     self.execute(sql, parameters)
pid-89470 thread-140735094956800 (MainThread)   File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/psycopg2/extras.py", line 223, in execute
pid-89470 thread-140735094956800 (MainThread)     return super(RealDictCursor, self).execute(query, vars)
pid-89470 thread-140735094956800 (MainThread) ProgrammingError: can't adapt type 'Team'
```

File: gratipay/models/team/__init__.py [hunk 1]
```diff
@@ -367,3 +379,35 @@ def load_image(self, size):
...
+    qs = request.line.uri.querystring
+
+    try:
+        team = Team.from_slug(slug)
+    except:
+        raise Response(400, 'bad slug')
+
+    if team is None:
+        # Try to redirect to a Participant.
+        from gratipay.models.participant import Participant # avoid circular import
```
Comment (dowski) [inline gratipay/models/team/__init__.py]: Open a new GH issue to refactor this so you don't have to lazy-import. Circular dependencies are a hassle.

**Comment by whit537** [comment]
Rebased, was 89202f8cdd7249526b1ee23485252b1ac838acdd.

**Comment by whit537** [comment]
Rebasing, determined from https://github.com/gratipay/gratipay.com/pull/4397#issuecomment-294472892 that 757b498c1c8995280e9a735b314cd23c43a78b68 and previous can be pushed off of here. The work starts at 8151f865f249f20ff3ab3e450bf08eb3ef90467b.

**Comment by whit537** [comment]
Squashed to ease rebase, was 89202f8cdd7249526b1ee23485252b1ac838acdd.

**Comment by whit537** [comment]
Not sure what I was thinking. Conflict resolution is effectively the same. Lessee here ...

**Comment by whit537** [comment]
Okay! Now to address review and todo ...

File: gratipay/models/package/__init__.py [hunk 2]
```diff
@@ -1,6 +1,10 @@
 # -*- coding: utf-8 -*-
```
Comment (whit537) [inline gratipay/models/package/__init__.py]: Hopefully da0c3da8760695008586347fd9d287b1a823136c helps?

File: gratipay/models/team/__init__.py (hunk 1 above)
Comment (whit537) [inline gratipay/models/team/__init__.py]: Done: https://github.com/gratipay/gratipay.com/issues/4421.

File: gratipay/exceptions.py [hunk 3]
```diff
@@ -87,3 +87,6 @@ def __str__(self):
         return "Negative balance not allowed in this context."
 
 class NotWhitelisted(Exception): pass
+
+
+class OutOfOptions(Exception): pass
```
Comment (whit537) [inline gratipay/exceptions.py]: Done in 371231898b18551858b6858d764c1e6c06c4472e.

**Comment by whit537** [comment]
Alright @dowski, another one on deck!

File: gratipay/models/package/__init__.py [hunk 4]
```diff
@@ -78,7 +95,9 @@ def ensure_team(self, cursor, owner):
                                )
             cursor.run('INSERT INTO teams_to_packages (team_id, package_id) '
                        'VALUES (%s, %s)', (team.id, self.id))
-        return team
+            break
+        else:
+            raise OutOfOptions()
```
Comment (dowski) [inline gratipay/models/package/__init__.py]: This code will never run, unless UUID4 is broken. I'd put my money on UUID4 and remove the else and the exception.

Comment (whit537) [inline gratipay/models/package/__init__.py]: Done in 132eceb9faede40b75bb7261ea513069acadfee3.

**Pull request merged**